<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.6.3" />
<title>polished_scripts.STAR_alignment API documentation</title>
<meta name="description" content="&lt;h3&gt;Methods for running RNA alignment with STAR, including index generation.&lt;/h3&gt; â€¦" />
<link href='https://cdnjs.cloudflare.com/ajax/libs/normalize/8.0.0/normalize.min.css' rel='stylesheet'>
<link href='https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/8.0.0/sanitize.min.css' rel='stylesheet'>
<link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" rel="stylesheet">
<style>.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{font-weight:bold}#index h4 + ul{margin-bottom:.6em}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase;cursor:pointer}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>polished_scripts.STAR_alignment</code></h1>
</header>
<section id="section-intro">
<h3>Methods for running RNA alignment with STAR, including index generation.</h3>
<p>See <a href="http://labshare.cshl.edu/shares/gingeraslab/www-data/dobin/STAR/STAR.posix/doc/STARmanual.pdf"><u>here</u></a>
for resources on STAR (especially helpful was section 2.2 for determining which <code>.GFF</code> files are compatible).</p>
<p><strong>NB:</strong> Annotations used for alignment were from <a href="https://www.gencodegenes.org/human/"><u>GENCODE</u></a>,
which uses the UCSC convention labelling chromosomes like "chr1".</p>
<p>The default index used by this script can be found at:
<code>/n/data1/hms/dbmi/park/lawrence/bin/RNA_Seq/STAR_INDEX_GRCh38_GENCODE_annotations</code>.</p>
<p>This index was generated using the reference genome from:
<code>/n/data1/hms/dbmi/park/lawrence/reference_data/GRCh38.primary_assembly.genome_gencode.fa</code>
and the annotations file here: <code>/n/data1/hms/dbmi/park/lawrence/reference_data/gencode.v31.annotation.gff3</code>.</p>
<p>The full command to generate the STAR index was as follows:</p>
<pre><code>module load gcc/6.2.0 star/2.7.0f.lua
STAR --runMode genomeGenerate 
    --runThreadN 16 
    --genomeDir /n/data1/hms/dbmi/park/lawrence/bin/RNA_Seq/STAR_INDEX_GRCh38_GENCODE_annotations 
    --genomeFastaFiles /n/data1/hms/dbmi/park/lawrence/reference_data/GRCh38.primary_assembly.genome_gencode.fa 
    --sjdbGTFfile /n/data1/hms/dbmi/park/lawrence/reference_data/gencode.v31.annotation.gff3 
    --sjdbOverhang 100 
    --sjdbGTFtagExonParentTranscript Parent
</code></pre>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">&#34;&#34;&#34;&lt;h3&gt;Methods for running RNA alignment with STAR, including index generation.&lt;/h3&gt;

See &lt;a href=&#34;http://labshare.cshl.edu/shares/gingeraslab/www-data/dobin/STAR/STAR.posix/doc/STARmanual.pdf&#34;&gt;&lt;u&gt;here&lt;/u&gt;&lt;/a&gt;
for resources on STAR (especially helpful was section 2.2 for determining which `.GFF` files are compatible).

&lt;strong&gt;NB:&lt;/strong&gt; Annotations used for alignment were from &lt;a href=&#34;https://www.gencodegenes.org/human/&#34;&gt;&lt;u&gt;GENCODE&lt;/u&gt;&lt;/a&gt;, 
which uses the UCSC convention labelling chromosomes like &#34;chr1&#34;.

The default index used by this script can be found at: 
`/n/data1/hms/dbmi/park/lawrence/bin/RNA_Seq/STAR_INDEX_GRCh38_GENCODE_annotations`.

This index was generated using the reference genome from: 
`/n/data1/hms/dbmi/park/lawrence/reference_data/GRCh38.primary_assembly.genome_gencode.fa` 
and the annotations file here: `/n/data1/hms/dbmi/park/lawrence/reference_data/gencode.v31.annotation.gff3`.

The full command to generate the STAR index was as follows:

```
module load gcc/6.2.0 star/2.7.0f.lua
STAR --runMode genomeGenerate \
\n    --runThreadN 16 \
\n    --genomeDir /n/data1/hms/dbmi/park/lawrence/bin/RNA_Seq/STAR_INDEX_GRCh38_GENCODE_annotations \
\n    --genomeFastaFiles /n/data1/hms/dbmi/park/lawrence/reference_data/GRCh38.primary_assembly.genome_gencode.fa \
\n    --sjdbGTFfile /n/data1/hms/dbmi/park/lawrence/reference_data/gencode.v31.annotation.gff3 \
\n    --sjdbOverhang 100 \
\n    --sjdbGTFtagExonParentTranscript Parent
```
&#34;&#34;&#34;

import os
import glob
import argparse
import re

from polished_scripts.helpers import _split_all_ext, _handle_directory_cleaning

# CONSTANTS
ACCESSION_REGEX = re.compile(&#34;(SRR[0-9]+)&#34;)
INPUT_EXTENSION_REGEX = &#34;*.fastq*&#34;
    
def parse_args():
    &#34;&#34;&#34;Uses argparse to enable user to customize script functionality, returns a dictionary object.&#34;&#34;&#34;
    
    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument(&#34;-input_dir&#34;, &#34;--input_directory&#34;, default=None, help=&#34;Files with SRA Accessions to download, one accession per line&#34;)
    parser.add_argument(&#34;-index_dir&#34;, &#34;--index_directory&#34;, default=&#34;/n/data1/hms/dbmi/park/lawrence/bin/RNA_Seq/STAR_INDEX_GRCh38_GENCODE_annotations/&#34;, help=&#34;Optional directory with STAR indices, otherwise will use existing&#34;)
    parser.add_argument(&#34;-out&#34;, &#34;--output_directory&#34;, default=&#34;./&#34;, help=&#34;Output directory for aligned reads&#34;)
    parser.add_argument(&#34;-n&#34;, &#34;--num_cores&#34;, default=&#34;16&#34;, help=&#34;slurm job submission option&#34;)
    parser.add_argument(&#34;-t&#34;, &#34;--runtime&#34;, default=&#34;0-2:00:00&#34;, help=&#34;slurm job submission option&#34;)
    parser.add_argument(&#34;-p&#34;, &#34;--queue&#34;, default=&#34;park&#34;, help=&#34;slurm job submission option&#34;)
    parser.add_argument(&#34;--mem_per_cpu&#34;, default=&#34;50G&#34;, help=&#34;slurm job submission option, NB: the filtering step requires a lot of memory because the mapper razers3 loads all the reads into memory&#34;)
    parser.add_argument(&#34;--mail_type&#34;, default=&#34;ALL&#34;, help=&#34;slurm job submission option&#34;)
    # parser.add_argument(&#34;--mail_user&#34;, default=&#34;alon_galor@hms.harvard.edu&#34;, help=&#34;slurm job submission option&#34;)
    parser.add_argument(&#34;--mail_user&#34;, default=&#34;lawrence_huang1@brown.edu&#34;, help=&#34;slurm job submission option&#34;)
    return vars(parser.parse_args())

def clean_arg_paths(args_dict):
    &#34;&#34;&#34;Replaces &#39;.&#39; in paths, modifying args_dict.&#34;&#34;&#34;
    
    # replaces paths starting with &#34;.&#34; for input/output directory
    input_dir = args_dict[&#34;input_directory&#34;]
    output_dir = args_dict[&#34;output_directory&#34;]
    index_dir = args_dict[&#34;index_directory&#34;]
    if input_dir is not None:
        args_dict[&#34;input_directory&#34;] = _handle_directory_cleaning(input_dir)
    if index_dir is not None:
        args_dict[&#34;index_directory&#34;] = _handle_directory_cleaning(index_dir)
    if output_dir is not None:
        args_dict[&#34;output_directory&#34;] = _handle_directory_cleaning(output_dir)
    
def return_slurm_command(args_dict):
    &#34;&#34;&#34;Returns slurm command given args_dict provided as string.&#34;&#34;&#34;
    
    slurm_command = &#34;#!/bin/bash\n&#34; + \
                &#34;#SBATCH -n &#34; + args_dict[&#34;num_cores&#34;] + &#34;\n&#34; + \
                &#34;#SBATCH -t &#34; + args_dict[&#34;runtime&#34;] + &#34;\n&#34; + \
                &#34;#SBATCH -p &#34; + args_dict[&#34;queue&#34;] + &#34;\n&#34; + \
                &#34;#SBATCH --mem-per-cpu=&#34; + args_dict[&#34;mem_per_cpu&#34;] + &#34;\n&#34; + \
                &#34;#SBATCH --mail-type=&#34; + args_dict[&#34;mail_type&#34;] + &#34;\n&#34; + \
                &#34;#SBATCH --mail-user=&#34; + args_dict[&#34;mail_user&#34;] + &#34;\n&#34;
    if args_dict[&#34;queue&#34;] in [&#34;park&#34;, &#34;priopark&#34;]:
        slurm_command += &#34;#SBATCH --account=park_contrib&#34; + &#34;\n&#34;

    return slurm_command

def get_input_files(input_dir):
    &#34;&#34;&#34;Returns list of input FASTQ&#39;s from input directory.&#34;&#34;&#34;
    return [input_file for input_file in glob.glob(os.path.join(input_dir, INPUT_EXTENSION_REGEX))]

def run_alignment(slurm_command, input_files, index_dir, output_dir, numThreads):
    &#34;&#34;&#34;Creates and submits alignment scripts.&#34;&#34;&#34;
    os.makedirs(os.path.join(output_dir, &#34;.sh&#34;), exist_ok=True)
    for input_file in input_files:
        # search for accession in file name
        run_name_match = ACCESSION_REGEX.search(os.path.basename(input_file))
        if run_name_match:
            run_name = run_name_match.group(1)
            main_command = f&#34;module load gcc/6.2.0 star/2.7.0f; STAR --genomeDir {index_dir} &#34; + \
            f&#34;--runThreadN {numThreads} --readFilesIn {input_file} --outFileNamePrefix {run_name}&#34;
            
            script_name = os.path.join(output_dir, &#34;.sh&#34;, run_name + &#34;_STAR_align.sh&#34;)
            with open(script_name, &#34;w+&#34;) as fp:
                fp.write(slurm_command + main_command)
            os.system(&#34;chmod u+x &#34; + script_name)
            os.system(&#34;sbatch &#34; + script_name)

if __name__ == &#34;__main__&#34;:
    args_dict = parse_args()
    clean_arg_paths(args_dict)
    
    input_dir = args_dict[&#34;input_directory&#34;]
    index_dir = args_dict[&#34;index_directory&#34;]
    output_dir = args_dict[&#34;output_directory&#34;]
    
    if input_dir is not None: 
        slurm_command = return_slurm_command(args_dict)
        input_files = get_input_files(input_dir)
        
        run_alignment(slurm_command, input_files, index_dir, output_dir, args_dict[&#34;num_cores&#34;])
    else:
        print(&#34;No input directory specified.&#34;)</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="polished_scripts.STAR_alignment.clean_arg_paths"><code class="name flex">
<span>def <span class="ident">clean_arg_paths</span></span>(<span>args_dict)</span>
</code></dt>
<dd>
<section class="desc"><p>Replaces '.' in paths, modifying args_dict.</p></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def clean_arg_paths(args_dict):
    &#34;&#34;&#34;Replaces &#39;.&#39; in paths, modifying args_dict.&#34;&#34;&#34;
    
    # replaces paths starting with &#34;.&#34; for input/output directory
    input_dir = args_dict[&#34;input_directory&#34;]
    output_dir = args_dict[&#34;output_directory&#34;]
    index_dir = args_dict[&#34;index_directory&#34;]
    if input_dir is not None:
        args_dict[&#34;input_directory&#34;] = _handle_directory_cleaning(input_dir)
    if index_dir is not None:
        args_dict[&#34;index_directory&#34;] = _handle_directory_cleaning(index_dir)
    if output_dir is not None:
        args_dict[&#34;output_directory&#34;] = _handle_directory_cleaning(output_dir)</code></pre>
</details>
</dd>
<dt id="polished_scripts.STAR_alignment.get_input_files"><code class="name flex">
<span>def <span class="ident">get_input_files</span></span>(<span>input_dir)</span>
</code></dt>
<dd>
<section class="desc"><p>Returns list of input FASTQ's from input directory.</p></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def get_input_files(input_dir):
    &#34;&#34;&#34;Returns list of input FASTQ&#39;s from input directory.&#34;&#34;&#34;
    return [input_file for input_file in glob.glob(os.path.join(input_dir, INPUT_EXTENSION_REGEX))]</code></pre>
</details>
</dd>
<dt id="polished_scripts.STAR_alignment.parse_args"><code class="name flex">
<span>def <span class="ident">parse_args</span></span>(<span>)</span>
</code></dt>
<dd>
<section class="desc"><p>Uses argparse to enable user to customize script functionality, returns a dictionary object.</p></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def parse_args():
    &#34;&#34;&#34;Uses argparse to enable user to customize script functionality, returns a dictionary object.&#34;&#34;&#34;
    
    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument(&#34;-input_dir&#34;, &#34;--input_directory&#34;, default=None, help=&#34;Files with SRA Accessions to download, one accession per line&#34;)
    parser.add_argument(&#34;-index_dir&#34;, &#34;--index_directory&#34;, default=&#34;/n/data1/hms/dbmi/park/lawrence/bin/RNA_Seq/STAR_INDEX_GRCh38_GENCODE_annotations/&#34;, help=&#34;Optional directory with STAR indices, otherwise will use existing&#34;)
    parser.add_argument(&#34;-out&#34;, &#34;--output_directory&#34;, default=&#34;./&#34;, help=&#34;Output directory for aligned reads&#34;)
    parser.add_argument(&#34;-n&#34;, &#34;--num_cores&#34;, default=&#34;16&#34;, help=&#34;slurm job submission option&#34;)
    parser.add_argument(&#34;-t&#34;, &#34;--runtime&#34;, default=&#34;0-2:00:00&#34;, help=&#34;slurm job submission option&#34;)
    parser.add_argument(&#34;-p&#34;, &#34;--queue&#34;, default=&#34;park&#34;, help=&#34;slurm job submission option&#34;)
    parser.add_argument(&#34;--mem_per_cpu&#34;, default=&#34;50G&#34;, help=&#34;slurm job submission option, NB: the filtering step requires a lot of memory because the mapper razers3 loads all the reads into memory&#34;)
    parser.add_argument(&#34;--mail_type&#34;, default=&#34;ALL&#34;, help=&#34;slurm job submission option&#34;)
    # parser.add_argument(&#34;--mail_user&#34;, default=&#34;alon_galor@hms.harvard.edu&#34;, help=&#34;slurm job submission option&#34;)
    parser.add_argument(&#34;--mail_user&#34;, default=&#34;lawrence_huang1@brown.edu&#34;, help=&#34;slurm job submission option&#34;)
    return vars(parser.parse_args())</code></pre>
</details>
</dd>
<dt id="polished_scripts.STAR_alignment.return_slurm_command"><code class="name flex">
<span>def <span class="ident">return_slurm_command</span></span>(<span>args_dict)</span>
</code></dt>
<dd>
<section class="desc"><p>Returns slurm command given args_dict provided as string.</p></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def return_slurm_command(args_dict):
    &#34;&#34;&#34;Returns slurm command given args_dict provided as string.&#34;&#34;&#34;
    
    slurm_command = &#34;#!/bin/bash\n&#34; + \
                &#34;#SBATCH -n &#34; + args_dict[&#34;num_cores&#34;] + &#34;\n&#34; + \
                &#34;#SBATCH -t &#34; + args_dict[&#34;runtime&#34;] + &#34;\n&#34; + \
                &#34;#SBATCH -p &#34; + args_dict[&#34;queue&#34;] + &#34;\n&#34; + \
                &#34;#SBATCH --mem-per-cpu=&#34; + args_dict[&#34;mem_per_cpu&#34;] + &#34;\n&#34; + \
                &#34;#SBATCH --mail-type=&#34; + args_dict[&#34;mail_type&#34;] + &#34;\n&#34; + \
                &#34;#SBATCH --mail-user=&#34; + args_dict[&#34;mail_user&#34;] + &#34;\n&#34;
    if args_dict[&#34;queue&#34;] in [&#34;park&#34;, &#34;priopark&#34;]:
        slurm_command += &#34;#SBATCH --account=park_contrib&#34; + &#34;\n&#34;

    return slurm_command</code></pre>
</details>
</dd>
<dt id="polished_scripts.STAR_alignment.run_alignment"><code class="name flex">
<span>def <span class="ident">run_alignment</span></span>(<span>slurm_command, input_files, index_dir, output_dir, numThreads)</span>
</code></dt>
<dd>
<section class="desc"><p>Creates and submits alignment scripts.</p></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def run_alignment(slurm_command, input_files, index_dir, output_dir, numThreads):
    &#34;&#34;&#34;Creates and submits alignment scripts.&#34;&#34;&#34;
    os.makedirs(os.path.join(output_dir, &#34;.sh&#34;), exist_ok=True)
    for input_file in input_files:
        # search for accession in file name
        run_name_match = ACCESSION_REGEX.search(os.path.basename(input_file))
        if run_name_match:
            run_name = run_name_match.group(1)
            main_command = f&#34;module load gcc/6.2.0 star/2.7.0f; STAR --genomeDir {index_dir} &#34; + \
            f&#34;--runThreadN {numThreads} --readFilesIn {input_file} --outFileNamePrefix {run_name}&#34;
            
            script_name = os.path.join(output_dir, &#34;.sh&#34;, run_name + &#34;_STAR_align.sh&#34;)
            with open(script_name, &#34;w+&#34;) as fp:
                fp.write(slurm_command + main_command)
            os.system(&#34;chmod u+x &#34; + script_name)
            os.system(&#34;sbatch &#34; + script_name)</code></pre>
</details>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="polished_scripts" href="index.html">polished_scripts</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="polished_scripts.STAR_alignment.clean_arg_paths" href="#polished_scripts.STAR_alignment.clean_arg_paths">clean_arg_paths</a></code></li>
<li><code><a title="polished_scripts.STAR_alignment.get_input_files" href="#polished_scripts.STAR_alignment.get_input_files">get_input_files</a></code></li>
<li><code><a title="polished_scripts.STAR_alignment.parse_args" href="#polished_scripts.STAR_alignment.parse_args">parse_args</a></code></li>
<li><code><a title="polished_scripts.STAR_alignment.return_slurm_command" href="#polished_scripts.STAR_alignment.return_slurm_command">return_slurm_command</a></code></li>
<li><code><a title="polished_scripts.STAR_alignment.run_alignment" href="#polished_scripts.STAR_alignment.run_alignment">run_alignment</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.6.3</a>.</p>
</footer>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
<script>hljs.initHighlightingOnLoad()</script>
</body>
</html>