<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.6.3" />
<title>polished_scripts.Optitype_HLA_Typing API documentation</title>
<meta name="description" content="&lt;h3&gt;Determines HLA Class I alleles from &lt;strong&gt;&lt;u&gt;paired end&lt;/u&gt;&lt;/strong&gt;
FASTQ&#39;s using the Optitype tool.&lt;/h3&gt; â€¦" />
<link href='https://cdnjs.cloudflare.com/ajax/libs/normalize/8.0.0/normalize.min.css' rel='stylesheet'>
<link href='https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/8.0.0/sanitize.min.css' rel='stylesheet'>
<link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" rel="stylesheet">
<style>.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{font-weight:bold}#index h4 + ul{margin-bottom:.6em}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase;cursor:pointer}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>polished_scripts.Optitype_HLA_Typing</code></h1>
</header>
<section id="section-intro">
<h3>Determines HLA Class I alleles from <strong><u>paired end</u></strong>
FASTQ's using the Optitype tool.</h3>
<p>Meant to be run as a command line script. Run with <code>--help</code> flag for options.
Assumes that input FASTQs have the format SRR{accession}_{#} where {#} is
either 1 or 2, indicating the paired read.</p>
<p>Each FASTQ undergoes two stages of processing:</p>
<ol>
<li>
<p>Filtering of reads that map to the HLA regions (this step is technically
optional but enables faster reprocessing if <code>optitype</code> needs to be run again)</p>
</li>
<li>
<p>Determination of HLA class I alleles from the filtered reads</p>
</li>
</ol>
<p><strong>NB: paired reads are filtered and HLA typed separately. A script
downstream in the pipeline will handle the pairing together of paired reads.</strong></p>
<p><strong>NB:</strong> Requires installation of the <code>optitype</code> <code>conda</code> environment
(see <a href="https://bioconda.github.io/recipes/optitype/README.html"><u>link</u></a>).</p>
<p>Original Github documentation <a href="https://github.com/FRED-2/OptiType">
<u>here</u></a> (contains explanation of filtering step, specific dependencies,
etc.). It's also possible to install from source, but the <code>conda</code> environment is
one simple command to install (some of the dependencies were slightly different
between the original Optitype directions and what was in the <code>conda</code> install but
this did not affect functionality).</p>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">&#34;&#34;&#34;
&lt;h3&gt;Determines HLA Class I alleles from &lt;strong&gt;&lt;u&gt;paired end&lt;/u&gt;&lt;/strong&gt; 
FASTQ&#39;s using the Optitype tool.&lt;/h3&gt;

Meant to be run as a command line script. Run with `--help` flag for options. 
Assumes that input FASTQs have the format SRR{accession}_{#} where {#} is 
either 1 or 2, indicating the paired read.

Each FASTQ undergoes two stages of processing:

1. Filtering of reads that map to the HLA regions (this step is technically 
optional but enables faster reprocessing if `optitype` needs to be run again)

2. Determination of HLA class I alleles from the filtered reads

&lt;strong&gt;NB: paired reads are filtered and HLA typed separately. A script 
downstream in the pipeline will handle the pairing together of paired reads.&lt;/strong&gt;

&lt;strong&gt;NB:&lt;/strong&gt; Requires installation of the `optitype` `conda` environment
(see &lt;a href=&#34;https://bioconda.github.io/recipes/optitype/README.html&#34;&gt;&lt;u&gt;link&lt;/u&gt;&lt;/a&gt;).

Original Github documentation &lt;a href=&#34;https://github.com/FRED-2/OptiType&#34;&gt;
&lt;u&gt;here&lt;/u&gt;&lt;/a&gt; (contains explanation of filtering step, specific dependencies, 
etc.). It&#39;s also possible to install from source, but the `conda` environment is
one simple command to install (some of the dependencies were slightly different 
between the original Optitype directions and what was in the `conda` install but
this did not affect functionality).
&#34;&#34;&#34;

import os
import glob
import argparse
import re

from .helpers import _split_all_ext, _handle_directory_cleaning

# CONSTANTS
OPTITYPE_CONFIG_PATH = &#34;/n/data1/hms/dbmi/park/alon/command_line_tools/Optitype/config.ini&#34;
PAIRED_END_ACCESSION = re.compile(&#34;(SRR[0-9]+)_([0-9]{1})&#34;)
INPUT_EXTENSION_REGEX = &#34;*.fastq*&#34;

# Pipeline methods
def generate_paired_reads_dict(input_dir):
    &#34;&#34;&#34;Associates paths of paired end reads together by accession, returns a dictionary of paths&#34;&#34;&#34;
    paired_reads_path_dict = {}
    acc_to_paths_dict = {}
    
    for file_path in glob.glob(os.path.join(input_dir, INPUT_EXTENSION_REGEX)):
        accession_match = PAIRED_END_ACCESSION.search(file_path)
        
        read_num = None
        if accession_match:
            accession = accession_match.group(1)
            read_num = int(accession_match.group(2))
        
            # each accession should point to two file paths in the list, 
            # indicating paired reads one and two
            if accession not in acc_to_paths_dict:
                acc_to_paths_dict[accession] = [None, None]
            acc_to_paths_dict[accession][read_num-1] = file_path
            
    for accession in acc_to_paths_dict:
        collected_paths = acc_to_paths_dict[accession]
        if None not in collected_paths:
            paired_reads_path_dict[collected_paths[0]] = collected_paths[1]
    
    return paired_reads_path_dict

def parse_args():
    &#34;&#34;&#34;Uses argparse to enable user to customize script functionality, returns a dictionary object.&#34;&#34;&#34;
    
    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument(&#34;-in_dir&#34;, &#34;--input_directory&#34;, default=None, help=&#34;Files with SRA Accessions to download, one accession per line&#34;)
    parser.add_argument(&#34;-out&#34;, &#34;--output_directory&#34;, default=&#34;./&#34;, help=&#34;Optional output directory, default is current directory&#34;)
    parser.add_argument(&#34;-n&#34;, &#34;--num_cores&#34;, default=&#34;16&#34;, help=&#34;slurm job submission option&#34;)
    parser.add_argument(&#34;-t&#34;, &#34;--runtime&#34;, default=&#34;0-1:15:00&#34;, help=&#34;slurm job submission option&#34;)
    parser.add_argument(&#34;-p&#34;, &#34;--queue&#34;, default=&#34;park&#34;, help=&#34;slurm job submission option&#34;)
    parser.add_argument(&#34;--mem_per_cpu&#34;, default=&#34;80G&#34;, help=&#34;slurm job submission option, NB: the filtering step requires a lot of memory because the mapper razers3 loads all the reads into memory&#34;)
    parser.add_argument(&#34;--mail_type&#34;, default=&#34;ALL&#34;, help=&#34;slurm job submission option&#34;)
    parser.add_argument(&#34;--mail_user&#34;, default=&#34;alon_galor@hms.harvard.edu&#34;, help=&#34;slurm job submission option&#34;)
    return vars(parser.parse_args())

def clean_arg_paths(args_dict):
    &#34;&#34;&#34;Replaces &#39;.&#39; in paths, modifying args_dict.&#34;&#34;&#34;
    
    # replaces paths starting with &#34;.&#34; for input/output directory
    input_dir = args_dict[&#34;input_directory&#34;]
    output_dir = args_dict[&#34;output_directory&#34;]
    if input_dir is not None:
        args_dict[&#34;input_directory&#34;] = _handle_directory_cleaning(input_dir)
    if output_dir is not None:
        args_dict[&#34;output_directory&#34;] = _handle_directory_cleaning(output_dir)

def return_slurm_command(args_dict):
    &#34;&#34;&#34;Returns slurm command given args_dict provided as string.&#34;&#34;&#34;
    
    slurm_command = &#34;#!/bin/bash\n&#34; + \
                &#34;#SBATCH -n &#34; + args_dict[&#34;num_cores&#34;] + &#34;\n&#34; + \
                &#34;#SBATCH -t &#34; + args_dict[&#34;runtime&#34;] + &#34;\n&#34; + \
                &#34;#SBATCH -p &#34; + args_dict[&#34;queue&#34;] + &#34;\n&#34; + \
                &#34;#SBATCH --mem-per-cpu=&#34; + args_dict[&#34;mem_per_cpu&#34;] + &#34;\n&#34; + \
                &#34;#SBATCH --mail-type=&#34; + args_dict[&#34;mail_type&#34;] + &#34;\n&#34; + \
                &#34;#SBATCH --mail-user=&#34; + args_dict[&#34;mail_user&#34;] + &#34;\n&#34;
    if args_dict[&#34;queue&#34;] in [&#34;park&#34;, &#34;priopark&#34;]:
        slurm_command += &#34;#SBATCH --account=park_contrib&#34; + &#34;\n&#34;
    slurm_command += &#34;module load conda2/4.2.13.lua\n&#34; + \
    &#34;source activate /home/ag457/anaconda3/envs/optitype\n&#34;

    return slurm_command

def return_filter_command(output_dir, input_file):
    &#34;&#34;&#34;
    Returns command to filters HLA reads into an intermediate directory contained in the output directory for faster repeated optitype runs.
    
    See `optitype` Github for explanation of this step. The filtered files are appended with `_fished.{original extension of input_file}`.
    &#34;&#34;&#34;
    output_base_name = os.path.join(output_dir, _split_all_ext(os.path.basename(input_file))[0])
    input_extension = _split_all_ext(os.path.basename(input_file))[1]
    
    filter_command = &#34;echo \&#34;Starting razers3 for file {base_name} at $(date +\&#34;%T\&#34;)\&#34;\n&#34; + \
    &#34;razers3 -i 95 -m 1 -dr 0 -o {base_name}_fished.bam /n/data1/hms/dbmi/park/alon/command_line_tools/Optitype/data/hla_reference_dna.fasta {input_file}\n&#34; + \
    &#34;echo \&#34;Finished razers3 for file {base_name}, starting bam2fq at $(date +\&#34;%T\&#34;)\&#34;\n&#34; + \
    &#34;samtools bam2fq {base_name}_fished.bam &gt; {base_name}_fished{input_extension}\n&#34; + \
    &#34;rm {base_name}_fished.bam\n&#34; + \
    &#34;echo \&#34;Finished filtering for file {base_name} at $(date +\&#34;%T\&#34;)\&#34;\n&#34;
    filter_command = filter_command.format(base_name=output_base_name, input_file=input_file, input_extension=input_extension)
    
    filtered_output_path = &#34;{base_name}_fished{input_extension}&#34;.format(base_name=output_base_name, input_extension=input_extension)
    return (filter_command, filtered_output_path)

def return_optitype_command(input_fished_file_1, input_fished_file_2, optitype_out_dir):
    &#34;&#34;&#34;Returns command to run optitype on filtered HLA paired reads file.&#34;&#34;&#34;
    
    optitype_command = &#34;python /home/ag457/anaconda3/envs/optitype/bin/OptiTypePipeline.py -i {input_fished_file_1} {input_fished_file_2} --dna -c {CONFIG} --verbose --outdir {out_dir}\n&#34;.format(input_fished_file_1=input_fished_file_1, input_fished_file_2=input_fished_file_2, CONFIG=OPTITYPE_CONFIG_PATH, out_dir=optitype_out_dir)
    
    return optitype_command

def make_optitype_sh(output_dir, complete_command, input_accession):
    &#34;&#34;&#34;Generates shell script for given paired reads and submits to O2&#34;&#34;&#34;
    SCRIPTS_DIR = os.path.join(output_dir, &#34;.optitype_sh&#34;)
    os.makedirs(SCRIPTS_DIR, exist_ok = True)
    script_path = os.path.join(SCRIPTS_DIR, input_accession + &#34;_optitype.sh&#34;)
    
    with open(script_path, &#34;w&#34;) as fp:
        fp.write(complete_command)
    # change permissions and submit to O2
    os.chmod(script_path, 0o700)
    os.system(&#34;sbatch &#34; + script_path)

def run_filter_optitype(paired_reads_path_dict, slurm_command, output_dir):
    &#34;&#34;&#34;Driver method to process all paired reads specified in the dictionary&#34;&#34;&#34;
    filtered_read_dir = os.path.join(output_dir, &#34;.HLA_FILTERED&#34;)
    os.makedirs(filtered_read_dir, exist_ok=True)
    
    for read_1_path in paired_reads_path_dict:
        read_2_path = paired_reads_path_dict[read_1_path]
        
        # generate filter commands for each paired read, which need to run
        # before HLA typing
        (filter_command_1, filtered_read_1_path) = return_filter_command(filtered_read_dir, read_1_path)
        (filter_command_2, filtered_read_2_path) = return_filter_command(filtered_read_dir, read_2_path)
        
        # create output directory by accession because the output of optitype
        # is a timestamped folder that loses accession information
        accession = PAIRED_END_ACCESSION.search(os.path.basename(read_1_path)).group(1)
        final_HLA_types_dir = os.path.join(output_dir, &#34;.HLA&#34;, accession)
        os.makedirs(final_HLA_types_dir, exist_ok=True)
        optitype_command = return_optitype_command(filtered_read_1_path, filtered_read_2_path, final_HLA_types_dir)
        
        complete_command = slurm_command + filter_command_1 + filter_command_2 + optitype_command
        make_optitype_sh(output_dir, complete_command, accession)
        

if __name__ == &#34;__main__&#34;:
    args_dict = parse_args()
    clean_arg_paths(args_dict)
    
    input_dir = args_dict[&#34;input_directory&#34;]
    output_dir = args_dict[&#34;output_directory&#34;]
    
    if input_dir is not None: 
        slurm_command = return_slurm_command(args_dict)
        paired_reads_path_dict = generate_paired_reads_dict(input_dir)
        run_filter_optitype(paired_reads_path_dict, slurm_command, output_dir)
    else:
        print(&#34;No input directory specified.&#34;)</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="polished_scripts.Optitype_HLA_Typing.clean_arg_paths"><code class="name flex">
<span>def <span class="ident">clean_arg_paths</span></span>(<span>args_dict)</span>
</code></dt>
<dd>
<section class="desc"><p>Replaces '.' in paths, modifying args_dict.</p></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def clean_arg_paths(args_dict):
    &#34;&#34;&#34;Replaces &#39;.&#39; in paths, modifying args_dict.&#34;&#34;&#34;
    
    # replaces paths starting with &#34;.&#34; for input/output directory
    input_dir = args_dict[&#34;input_directory&#34;]
    output_dir = args_dict[&#34;output_directory&#34;]
    if input_dir is not None:
        args_dict[&#34;input_directory&#34;] = _handle_directory_cleaning(input_dir)
    if output_dir is not None:
        args_dict[&#34;output_directory&#34;] = _handle_directory_cleaning(output_dir)</code></pre>
</details>
</dd>
<dt id="polished_scripts.Optitype_HLA_Typing.generate_paired_reads_dict"><code class="name flex">
<span>def <span class="ident">generate_paired_reads_dict</span></span>(<span>input_dir)</span>
</code></dt>
<dd>
<section class="desc"><p>Associates paths of paired end reads together by accession, returns a dictionary of paths</p></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def generate_paired_reads_dict(input_dir):
    &#34;&#34;&#34;Associates paths of paired end reads together by accession, returns a dictionary of paths&#34;&#34;&#34;
    paired_reads_path_dict = {}
    acc_to_paths_dict = {}
    
    for file_path in glob.glob(os.path.join(input_dir, INPUT_EXTENSION_REGEX)):
        accession_match = PAIRED_END_ACCESSION.search(file_path)
        
        read_num = None
        if accession_match:
            accession = accession_match.group(1)
            read_num = int(accession_match.group(2))
        
            # each accession should point to two file paths in the list, 
            # indicating paired reads one and two
            if accession not in acc_to_paths_dict:
                acc_to_paths_dict[accession] = [None, None]
            acc_to_paths_dict[accession][read_num-1] = file_path
            
    for accession in acc_to_paths_dict:
        collected_paths = acc_to_paths_dict[accession]
        if None not in collected_paths:
            paired_reads_path_dict[collected_paths[0]] = collected_paths[1]
    
    return paired_reads_path_dict</code></pre>
</details>
</dd>
<dt id="polished_scripts.Optitype_HLA_Typing.make_optitype_sh"><code class="name flex">
<span>def <span class="ident">make_optitype_sh</span></span>(<span>output_dir, complete_command, input_accession)</span>
</code></dt>
<dd>
<section class="desc"><p>Generates shell script for given paired reads and submits to O2</p></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def make_optitype_sh(output_dir, complete_command, input_accession):
    &#34;&#34;&#34;Generates shell script for given paired reads and submits to O2&#34;&#34;&#34;
    SCRIPTS_DIR = os.path.join(output_dir, &#34;.optitype_sh&#34;)
    os.makedirs(SCRIPTS_DIR, exist_ok = True)
    script_path = os.path.join(SCRIPTS_DIR, input_accession + &#34;_optitype.sh&#34;)
    
    with open(script_path, &#34;w&#34;) as fp:
        fp.write(complete_command)
    # change permissions and submit to O2
    os.chmod(script_path, 0o700)
    os.system(&#34;sbatch &#34; + script_path)</code></pre>
</details>
</dd>
<dt id="polished_scripts.Optitype_HLA_Typing.parse_args"><code class="name flex">
<span>def <span class="ident">parse_args</span></span>(<span>)</span>
</code></dt>
<dd>
<section class="desc"><p>Uses argparse to enable user to customize script functionality, returns a dictionary object.</p></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def parse_args():
    &#34;&#34;&#34;Uses argparse to enable user to customize script functionality, returns a dictionary object.&#34;&#34;&#34;
    
    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument(&#34;-in_dir&#34;, &#34;--input_directory&#34;, default=None, help=&#34;Files with SRA Accessions to download, one accession per line&#34;)
    parser.add_argument(&#34;-out&#34;, &#34;--output_directory&#34;, default=&#34;./&#34;, help=&#34;Optional output directory, default is current directory&#34;)
    parser.add_argument(&#34;-n&#34;, &#34;--num_cores&#34;, default=&#34;16&#34;, help=&#34;slurm job submission option&#34;)
    parser.add_argument(&#34;-t&#34;, &#34;--runtime&#34;, default=&#34;0-1:15:00&#34;, help=&#34;slurm job submission option&#34;)
    parser.add_argument(&#34;-p&#34;, &#34;--queue&#34;, default=&#34;park&#34;, help=&#34;slurm job submission option&#34;)
    parser.add_argument(&#34;--mem_per_cpu&#34;, default=&#34;80G&#34;, help=&#34;slurm job submission option, NB: the filtering step requires a lot of memory because the mapper razers3 loads all the reads into memory&#34;)
    parser.add_argument(&#34;--mail_type&#34;, default=&#34;ALL&#34;, help=&#34;slurm job submission option&#34;)
    parser.add_argument(&#34;--mail_user&#34;, default=&#34;alon_galor@hms.harvard.edu&#34;, help=&#34;slurm job submission option&#34;)
    return vars(parser.parse_args())</code></pre>
</details>
</dd>
<dt id="polished_scripts.Optitype_HLA_Typing.return_filter_command"><code class="name flex">
<span>def <span class="ident">return_filter_command</span></span>(<span>output_dir, input_file)</span>
</code></dt>
<dd>
<section class="desc"><p>Returns command to filters HLA reads into an intermediate directory contained in the output directory for faster repeated optitype runs.</p>
<p>See <code>optitype</code> Github for explanation of this step. The filtered files are appended with <code>_fished.{original extension of input_file}</code>.</p></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def return_filter_command(output_dir, input_file):
    &#34;&#34;&#34;
    Returns command to filters HLA reads into an intermediate directory contained in the output directory for faster repeated optitype runs.
    
    See `optitype` Github for explanation of this step. The filtered files are appended with `_fished.{original extension of input_file}`.
    &#34;&#34;&#34;
    output_base_name = os.path.join(output_dir, _split_all_ext(os.path.basename(input_file))[0])
    input_extension = _split_all_ext(os.path.basename(input_file))[1]
    
    filter_command = &#34;echo \&#34;Starting razers3 for file {base_name} at $(date +\&#34;%T\&#34;)\&#34;\n&#34; + \
    &#34;razers3 -i 95 -m 1 -dr 0 -o {base_name}_fished.bam /n/data1/hms/dbmi/park/alon/command_line_tools/Optitype/data/hla_reference_dna.fasta {input_file}\n&#34; + \
    &#34;echo \&#34;Finished razers3 for file {base_name}, starting bam2fq at $(date +\&#34;%T\&#34;)\&#34;\n&#34; + \
    &#34;samtools bam2fq {base_name}_fished.bam &gt; {base_name}_fished{input_extension}\n&#34; + \
    &#34;rm {base_name}_fished.bam\n&#34; + \
    &#34;echo \&#34;Finished filtering for file {base_name} at $(date +\&#34;%T\&#34;)\&#34;\n&#34;
    filter_command = filter_command.format(base_name=output_base_name, input_file=input_file, input_extension=input_extension)
    
    filtered_output_path = &#34;{base_name}_fished{input_extension}&#34;.format(base_name=output_base_name, input_extension=input_extension)
    return (filter_command, filtered_output_path)</code></pre>
</details>
</dd>
<dt id="polished_scripts.Optitype_HLA_Typing.return_optitype_command"><code class="name flex">
<span>def <span class="ident">return_optitype_command</span></span>(<span>input_fished_file_1, input_fished_file_2, optitype_out_dir)</span>
</code></dt>
<dd>
<section class="desc"><p>Returns command to run optitype on filtered HLA paired reads file.</p></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def return_optitype_command(input_fished_file_1, input_fished_file_2, optitype_out_dir):
    &#34;&#34;&#34;Returns command to run optitype on filtered HLA paired reads file.&#34;&#34;&#34;
    
    optitype_command = &#34;python /home/ag457/anaconda3/envs/optitype/bin/OptiTypePipeline.py -i {input_fished_file_1} {input_fished_file_2} --dna -c {CONFIG} --verbose --outdir {out_dir}\n&#34;.format(input_fished_file_1=input_fished_file_1, input_fished_file_2=input_fished_file_2, CONFIG=OPTITYPE_CONFIG_PATH, out_dir=optitype_out_dir)
    
    return optitype_command</code></pre>
</details>
</dd>
<dt id="polished_scripts.Optitype_HLA_Typing.return_slurm_command"><code class="name flex">
<span>def <span class="ident">return_slurm_command</span></span>(<span>args_dict)</span>
</code></dt>
<dd>
<section class="desc"><p>Returns slurm command given args_dict provided as string.</p></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def return_slurm_command(args_dict):
    &#34;&#34;&#34;Returns slurm command given args_dict provided as string.&#34;&#34;&#34;
    
    slurm_command = &#34;#!/bin/bash\n&#34; + \
                &#34;#SBATCH -n &#34; + args_dict[&#34;num_cores&#34;] + &#34;\n&#34; + \
                &#34;#SBATCH -t &#34; + args_dict[&#34;runtime&#34;] + &#34;\n&#34; + \
                &#34;#SBATCH -p &#34; + args_dict[&#34;queue&#34;] + &#34;\n&#34; + \
                &#34;#SBATCH --mem-per-cpu=&#34; + args_dict[&#34;mem_per_cpu&#34;] + &#34;\n&#34; + \
                &#34;#SBATCH --mail-type=&#34; + args_dict[&#34;mail_type&#34;] + &#34;\n&#34; + \
                &#34;#SBATCH --mail-user=&#34; + args_dict[&#34;mail_user&#34;] + &#34;\n&#34;
    if args_dict[&#34;queue&#34;] in [&#34;park&#34;, &#34;priopark&#34;]:
        slurm_command += &#34;#SBATCH --account=park_contrib&#34; + &#34;\n&#34;
    slurm_command += &#34;module load conda2/4.2.13.lua\n&#34; + \
    &#34;source activate /home/ag457/anaconda3/envs/optitype\n&#34;

    return slurm_command</code></pre>
</details>
</dd>
<dt id="polished_scripts.Optitype_HLA_Typing.run_filter_optitype"><code class="name flex">
<span>def <span class="ident">run_filter_optitype</span></span>(<span>paired_reads_path_dict, slurm_command, output_dir)</span>
</code></dt>
<dd>
<section class="desc"><p>Driver method to process all paired reads specified in the dictionary</p></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def run_filter_optitype(paired_reads_path_dict, slurm_command, output_dir):
    &#34;&#34;&#34;Driver method to process all paired reads specified in the dictionary&#34;&#34;&#34;
    filtered_read_dir = os.path.join(output_dir, &#34;.HLA_FILTERED&#34;)
    os.makedirs(filtered_read_dir, exist_ok=True)
    
    for read_1_path in paired_reads_path_dict:
        read_2_path = paired_reads_path_dict[read_1_path]
        
        # generate filter commands for each paired read, which need to run
        # before HLA typing
        (filter_command_1, filtered_read_1_path) = return_filter_command(filtered_read_dir, read_1_path)
        (filter_command_2, filtered_read_2_path) = return_filter_command(filtered_read_dir, read_2_path)
        
        # create output directory by accession because the output of optitype
        # is a timestamped folder that loses accession information
        accession = PAIRED_END_ACCESSION.search(os.path.basename(read_1_path)).group(1)
        final_HLA_types_dir = os.path.join(output_dir, &#34;.HLA&#34;, accession)
        os.makedirs(final_HLA_types_dir, exist_ok=True)
        optitype_command = return_optitype_command(filtered_read_1_path, filtered_read_2_path, final_HLA_types_dir)
        
        complete_command = slurm_command + filter_command_1 + filter_command_2 + optitype_command
        make_optitype_sh(output_dir, complete_command, accession)</code></pre>
</details>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="polished_scripts" href="index.html">polished_scripts</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="polished_scripts.Optitype_HLA_Typing.clean_arg_paths" href="#polished_scripts.Optitype_HLA_Typing.clean_arg_paths">clean_arg_paths</a></code></li>
<li><code><a title="polished_scripts.Optitype_HLA_Typing.generate_paired_reads_dict" href="#polished_scripts.Optitype_HLA_Typing.generate_paired_reads_dict">generate_paired_reads_dict</a></code></li>
<li><code><a title="polished_scripts.Optitype_HLA_Typing.make_optitype_sh" href="#polished_scripts.Optitype_HLA_Typing.make_optitype_sh">make_optitype_sh</a></code></li>
<li><code><a title="polished_scripts.Optitype_HLA_Typing.parse_args" href="#polished_scripts.Optitype_HLA_Typing.parse_args">parse_args</a></code></li>
<li><code><a title="polished_scripts.Optitype_HLA_Typing.return_filter_command" href="#polished_scripts.Optitype_HLA_Typing.return_filter_command">return_filter_command</a></code></li>
<li><code><a title="polished_scripts.Optitype_HLA_Typing.return_optitype_command" href="#polished_scripts.Optitype_HLA_Typing.return_optitype_command">return_optitype_command</a></code></li>
<li><code><a title="polished_scripts.Optitype_HLA_Typing.return_slurm_command" href="#polished_scripts.Optitype_HLA_Typing.return_slurm_command">return_slurm_command</a></code></li>
<li><code><a title="polished_scripts.Optitype_HLA_Typing.run_filter_optitype" href="#polished_scripts.Optitype_HLA_Typing.run_filter_optitype">run_filter_optitype</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.6.3</a>.</p>
</footer>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
<script>hljs.initHighlightingOnLoad()</script>
</body>
</html>